program/rust/src/validator.rs

```diff
@@ -128,3 +141,53 @@ pub fn aggregate_price(
             .to_bytes(),
     ])
 }
+
+/// Load a price account as read-only, returning `None` if it isn't a valid price account.
+fn checked_load_price_account(price_account_info: &[u8]) -> Option<&PriceAccount> {
+    check_price_account_header(price_account_info).ok()?;
+    Some(bytemuck::from_bytes::<PriceAccount>(
+        &price_account_info[0..size_of::<PriceAccount>()],
+    ))
+}
+
+/// Computes the stake caps for each publisher based on the oracle program accounts provided
+/// - `account_datas` - the account datas of the oracle program accounts
+/// - `timestamp` - the timestamp to include in the message
+/// - `m` - m is the cap per symbol, it gets split among all publishers of the symbol
+/// - `z` - when a symbol has less than `z` publishers, each publisher gets a cap of `m/z` (instead of `m/number_of_publishers`). This is to prevent a single publisher from getting a large cap when there are few publishers.
+///
+/// The stake cap for a publisher is computed as the sum of `m/min(z, number_of_publishers)` for all the symbols the publisher is publishing.
+pub fn compute_publisher_stake_caps<'a>(
+    account_datas: impl IntoIterator<Item = &'a [u8]>,
+    timestamp: i64,
+    m: u64,
+    z: u64,
+) -> Vec<u8> {
+    let mut publisher_caps: BTreeMap<Pubkey, u64> = BTreeMap::new(); // BTreeMap to ensure it will be sorted by publisher
+    for account in account_datas {
+        if let Some(price_account) = checked_load_price_account(account) {
+            let cap: u64 = m
+                .checked_div(max(u64::from(price_account.num_), z))
+                .unwrap_or(0);
+            for i in 0..(price_account.num_ as usize) {
+                publisher_caps
+                    .entry(price_account.comp_[i].pub_)
```

Comment:
This is pretty unlikely to matter but it's safer to use checked indexing here of check `num_` beforehand to avoid crashing the validator if `num_` somehow ends up incorrect.

Reply:
Done